specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1096 @@
+
+When the checkpoint is committed, the `SCA` in `/root/f0100` is responsible for aggregating the checkpoint from `/root/f0100/f0200` with those of other children of `/root/f0100`, and for generating a new checkpoint for `/root/f0100`, which is then propagated to its parent chain, `/root`. The commitment of checkpoints also triggers the execution and propagation of cross-net messages. As checkpoints flow up the chain, the `SCA` of each chain picks up these checkpoints and inspects them to propagate potential state changes triggered by messages included in the cross-net messages that have the `SCA`'s subnet as a destination subnet (see [Cross-net messages](#Cross-net-messages "Cross-net messages")).
+
+Every checkpoint for a subnet points to the previous checkpoint being committed to ensure their integrity.
+
+> Checkpoint template read-only function in SCA
+```go
+// Returns the checkpoint template to be populated for the current
+// checkpoint window. The current checkpoint window is computed by 
+// (currEpoch / CheckPeriod) * CheckPeriod. This determines the checkpoint
+// that needs to be populated and signed for the current window.
+CheckpointTemplate() Checkpoint
+```
+
+As shown on the next figure, the checkpointing protocol has two distinct stages:
+- __Checkpoint Window__: In this window, the `SCA` opens a checkpoint template and starts populating it with cross-net messages arriving to the subnet (and that need to be propagated further inside the checkpoint) and with the aggregated checkpoints committed from the current subnet's children. The checkpoint window for the checkpoint in epoch `n` starts at `n-CheckPeriod` and ends at epoch `n`.
+- __Signing Window__: The signing window is the time range reserved for validators of the subnet to populate the checkpoint template with the corresponding proof of the state of the checkpoint, to sign the checkpoint, and to submit it to the corresponding `SA` in the parent for committment. The signing window for the checkpoint in epoch `n` goes from `n` to `n+CheckPeriod`. Consequently, the signing window for the checkpoint of epoch `n-CheckPeriod` and the checkpoint window for the checkpoint of epoch `n` is run in parallel. The checkpoint template provided by the `SCA` has `CrossMsgs`, `PrevCheck`, and `Epoch` already populated, and validators only have to add the corresponding `Proof` and sign it.
+
+
+![](https://hackmd.io/_uploads/HkAodIuFc.png)
```

Comment:
Should we check in the source code for the drawing?

Reply:
We should switch to some figure generation language for this(as you suggested in some other comment). These figures come from this slide deck (so no source code available): https://docs.google.com/presentation/d/1w-1XpwPrsBxjd1OScsbhuOvUdXz2DyQHAeY7gCbpfj4/edit

Do you mind if we resolve this comment and consider this for the next iteration?

Reply:
Sure! I just wanted to suggest that if there were any source code available, it could be a good idea to make them available for anyone who wants to maintain them. The link you included should be good for that!